### PR TITLE
Fixed #28558 -- Simplified code to remove OGRIndexError

### PR DESCRIPTION
**Pull request by pope1ni**
Ticket [#28558](https://code.djangoproject.com/ticket/28558).

In addition there are some other minor related cleanups, a fix for an index bounds checking error and improvements to exception messages. I've kept things in separate commits for ease of reviewing.

Let me know if anything need to be split up into a different pull request or any extra tests need adding.

The added test which is the exact equivalent to the one added in 197b1878105504b5ac7e399e1f52a6093c88baa3 still seems to work if I revert 12809e160995eb617fe394c75e5b9f3211c056ff so I'm not sure if there is something else that has changed with exception handling in the template engine lookup code.

### CODE REVIEW

**Comment by pope1ni** [comment]
Ticket [#28525](https://code.djangoproject.com/ticket/28525), pull request #8963, is waiting on this to determine whether to document `OGRIndexError`.

File: django/contrib/gis/gdal/datasource.py [hunk 1]
```diff
@@ -89,12 +89,12 @@ def __getitem__(self, index):
         if isinstance(index, str):
             layer = capi.get_layer_by_name(self.ptr, force_bytes(index))
             if not layer:
-                raise IndexError('invalid OGR Layer name given: "%s"' % index)
+                raise IndexError('Invalid OGR Layer name given: %s.' % index)
```
Comment (pope1ni) [inline django/contrib/gis/gdal/datasource.py]: There was no test case for this and it looks as though this doesn't work as expected.

Test failure when attempting to add an assertion for an invalid name:

```
======================================================================
ERROR: test01_valid_shp (gis_tests.gdal_tests.test_ds.DataSourceTest)
Testing valid SHP Data Source files.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib64/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File ".../django/tests/gis_tests/gdal_tests/test_ds.py", line 91, in test01_valid_shp
    ds['invalid']
  File ".../django/django/contrib/gis/gdal/datasource.py", line 90, in __getitem__
    layer = capi.get_layer_by_name(self.ptr, force_bytes(index))
  File ".../django/django/contrib/gis/gdal/prototypes/errcheck.py", line 128, in check_pointer
    raise GDALException('Invalid pointer returned from "%s"' % func.__name__)
django.contrib.gis.gdal.error.GDALException: Invalid pointer returned from "OGR_DS_GetLayerByName"
```

File: tests/gis_tests/geos_tests/test_mutable_list.py [hunk 2]
```diff
@@ -419,10 +419,7 @@ def test_12_arithmetic(self):
         self.assertLess(ul, pl + [2], 'cmp')
         self.assertLessEqual(ul, pl + [2], 'cmp')
 
-        # Also works with a custom IndexError
         ul_longer = ul + [2]
```
[the remark this answers is not in the document]
Comment (pope1ni) [reply tests/gis_tests/geos_tests/test_mutable_list.py]: Based on the assertions around it, I think this can go.

**Comment by pope1ni** [comment]
There was no test case for `DataSource.__getitem__` with a invalid string and it looks as though this doesn't work as expected.

Test failure when attempting to add an assertion for an invalid name:

```
======================================================================
ERROR: test01_valid_shp (gis_tests.gdal_tests.test_ds.DataSourceTest)
Testing valid SHP Data Source files.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib64/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File ".../django/tests/gis_tests/gdal_tests/test_ds.py", line 91, in test01_valid_shp
    ds['invalid']
  File ".../django/django/contrib/gis/gdal/datasource.py", line 90, in __getitem__
    layer = capi.get_layer_by_name(self.ptr, force_bytes(index))
  File ".../django/django/contrib/gis/gdal/prototypes/errcheck.py", line 128, in check_pointer
    raise GDALException('Invalid pointer returned from "%s"' % func.__name__)
django.contrib.gis.gdal.error.GDALException: Invalid pointer returned from "OGR_DS_GetLayerByName"
```

This should be fixed separately...

**Comment by pope1ni** [comment]
I've added a commit that fixes `DataSource.__getitem__` for invalid layer names.

File: tests/gis_tests/gdal_tests/test_ds.py [hunk 3]
```diff
@@ -1,11 +1,11 @@
 import os
 import re
-import unittest
 
 from django.contrib.gis.gdal import (
     GDAL_VERSION, DataSource, Envelope, GDALException, OGRGeometry,
 )
 from django.contrib.gis.gdal.field import OFTInteger, OFTReal, OFTString
+from django.test import TestCase
```
Comment (timgraham) [inline tests/gis_tests/gdal_tests/test_ds.py]: Can you use `SimpleTestCase` or are these these tests making queries?

File: tests/gis_tests/gdal_tests/test_ds.py [hunk 4]
```diff
@@ -83,8 +83,13 @@ def test01_valid_shp(self):
...
 
             # Making sure indexing works
-            with self.assertRaises(IndexError):
-                ds[len(ds)]
+            msg = 'Index out of range when accessing layers in a datasource: %s.'
+            with self.assertRaisesMessage(IndexError, msg % len(ds)):
+                ds.__getitem__(len(ds))
+
+            msg = 'Invalid OGR layer name given: %s.'
+            with self.assertRaisesMessage(IndexError, msg % 'invalid'):
```
Comment (timgraham) [inline tests/gis_tests/gdal_tests/test_ds.py]: I don't see a need for string interpolation in cases like this.

**Comment by pope1ni** [comment]
Did you want a new ticket for the `errcheck` fix? Or just the separate pull request?

**Comment by timgraham** [comment]
A ticket doesn't hurt.

**Pull request merged**